src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcServerTransport.cs

```diff
@@ -0,0 +1,201 @@
+            }
+        }
+    }
+
+    internal sealed class WindowsPipeServerTransport : IpcServerTransport
+    {
+        private const string PipePrefix = @"\\.\pipe\";
+
+        private NamedPipeServerStream _stream;
+
+        private readonly CancellationTokenSource _cancellation = new CancellationTokenSource();
+        private readonly string _pipeName;
+        private readonly int _maxInstances;
+
+        public WindowsPipeServerTransport(string pipeName, int maxInstances)
+        {
+            _maxInstances = maxInstances;
+            _pipeName = pipeName.StartsWith(PipePrefix) ? pipeName.Substring(PipePrefix.Length) : pipeName;
+            CreateNewPipeServer();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _cancellation.Cancel();
+
+                _stream.Dispose();
+
+                _cancellation.Dispose();
+            }
+        }
+
+        public override async Task<Stream> AcceptAsync(CancellationToken token)
+        {
+            VerifyNotDisposed();
+
+            using var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(token, _cancellation.Token);
+
+            NamedPipeServerStream connectedStream;
+            try
+            {
+                await _stream.WaitForConnectionAsync(linkedSource.Token).ConfigureAwait(false);
+
+                connectedStream = _stream;
+            }
+            finally
+            {
+                if (!_cancellation.IsCancellationRequested)
+                {
+                    CreateNewPipeServer();
+                }
+            }
+            return connectedStream;
+        }
+
+        private void CreateNewPipeServer()
+        {
+            _stream = new NamedPipeServerStream(
+                _pipeName,
+                PipeDirection.InOut,
+                _maxInstances,
+                PipeTransmissionMode.Byte,
+                PipeOptions.Asynchronous);
+        }
+    }
+
+    internal sealed class UnixDomainSocketServerTransport : IpcServerTransport
+    {
+        private readonly CancellationTokenSource _cancellation = new CancellationTokenSource();
+        private readonly int _backlog;
+        private readonly string _path;
+
+        private UnixDomainSocket _socket;
+
+        public UnixDomainSocketServerTransport(string path, int backlog)
+        {
+            _backlog = backlog;
+            _path = path;
+
+            CreateNewSocketServer();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _cancellation.Cancel();
+
+                try
+                {
+                    _socket.Shutdown(SocketShutdown.Both);
+                }
+                catch { }
+                finally
+                {
+                    _socket.Close(0);
+                }
+                _socket.Dispose();
```

Comment:
One thing to note: I found that on macOS, there is a race between `unlink`ing and `close`ing the socket for the reverse server (see dotnet/runtime#40225 for details).  TL;DR - if you don't `unlink` (or `File.Delete`) the Unix domain socket file before `close`ing the socket, you can end up with the runtime successfully connecting to the server descpite `close`ing the bound socket.  For some reason that results in a valid connection that the kernel doesn't close when you call `close` on the socket.  Right now, you have the file being deleted in the `UnixDomainSocket.Dispose` method, which is being called second here.

Reply:
That's going to be tricky to implement because the only thing that can be override from Socket is the Dispose method. I can add another Close* method implementation that all our code calls that deletes the file first and then calls the underlying Close. Maybe shadow implement the Socket.Close methods and have them Debug.Assert that the file wasn't specified so we can catch new callers in the future.

Reply:
This might actually not be an issue for the C# Socket class and how I'm deleting the socket file in the Dispose(bool) method because the Close methods effectively just set a timeout internal to the Socket class and then call Dispose. The Dispose method calls Dispose(bool), so my implementation gets called first and then it calls into the Socket.Dispose(bool) method which does the closing and disposing of its resources.

Reply:
So long as the Unix domain socket file is deleted before you `close` the socket bound to that address you should be okay.